src/cdomains/intDomain.ml

```diff
@@ -349,6 +349,446 @@ struct
+
+  let equal_to i = function
+    | None -> failwith "unsupported: equal_to with bottom"
+    | Some s ->
+      if not (SS.exists (fun (a, b) -> a <= i && i <= b) s) then `Neq else
+        match SS.cardinal s with
+          | 1 ->
+            let a, b = SS.min_elt s in
+            if a = b && b = i then `Eq else `Top
+          | _ -> `Top
+
+  let to_bool = function
+    | None -> None
+    | x when equal x zero -> Some false
+    | x -> if leq zero x then None else Some true
+
+  let to_bool_interval x =
+    match x with
+    | None -> None
+    | _ when equal x zero -> x
+    | _ -> if leq zero x then top_bool else one
+
+  let of_bool = function
+    | true -> Some (SS.add (1L, 1L) SS.empty)
+    | false -> Some (SS.add (0L, 0L) SS.empty)
+
+  let is_bool x =
+    x <> None && not (leq zero x) || equal x zero
+
+  let to_excl_list _ = None
+  let of_excl_list t _ = top ()
+  let is_excl_list _ = false
+
+  let of_interval (x,y) = norm @@ Some (x,y)
+  let of_int x = of_interval (x,x)
+  let zero = Some (SS.add (0L, 0L) SS.empty)
+  let one = Some (SS.add (1L, 1L) SS.empty)
+  let top_bool = Some (SS.add (0L, 1L) SS.empty)
+
+  let starting n = norm @@ Some (n, max_int)
+  let ending   n = norm @@ Some (min_int, n)
+  let maximal = function
+    | None -> None
+    | Some s ->
+      let max_elt = SS.max_elt s in
+      Some ((fun (x, y) -> y) max_elt)
+  let minimal = function
+    | None -> None
+    | Some s ->
+      let min_elt = SS.min_elt s in
+      Some ((fun (x, y) -> x) min_elt)
+
+  (* takes a sorted list of intervals, iterates through them and merges two adjacent if they intersect*)
+  let rec merge_intersect ?(acc = []) l =
+    match l with
+    | (x1, x2) :: xs when Int64.compare x1 x2 > 0 ->
+      merge_intersect ~acc:acc xs
+    | (x1, x2) :: (y1, y2) :: xs ->
+      if Int64.compare x2 (Int64.sub y1 1L) >= 0 then
+        merge_intersect ~acc:acc ((x1, max x2 y2) :: xs)
+      else merge_intersect ~acc:((x1, x2) :: acc) ((y1, y2) :: xs)
+    | [(x1, x2)] -> merge_intersect ~acc:((x1, x2) :: acc) []
+    | [] -> acc
+
+  (* recursive template for evaluating arithmetic operations which are performed pairwise
+     for each interval from both set operands *)
+  let rec arithm ?(acc1 = []) ?(acc2 = []) f l1 l2 =
+    if l1 = [] && l2 = [] && acc1 = [] && acc2 = [] then failwith "error: empty entry"
+    else
+      match l1, l2, acc2 with
+      | _, [], _::_ ->
+        let list_merged = merge_intersect (List.sort IntPair.compare acc2) in
+        if list_merged = [] then None else Some (SS.of_list list_merged)
+      | x1 :: xs1, x2 :: xs2, _ ->
+        let (y1, y2) = f x1 x2 in
+        if Int64.compare y1 min_int = 0 && Int64.compare y2 max_int = 0 then top ()
+          else arithm ~acc1:(x1 :: acc1) ~acc2:((y1,y2) :: acc2) f xs1 l2
+      | [], x2 :: xs2, _ -> arithm ~acc1:[] ~acc2:acc2 f acc1 xs2
+      | [], [], [] | _ :: _, [], [] -> failwith "error: empty entry"
+
+  let add_interval (x1, x2) (y1, y2) =
```

Comment:
For all these `_interval` you should be able to use the things that the `Interval32` domain provides.